src/Microsoft.Identity.Web/Policy/ScopeAuthorizationHandler.cs

```diff
@@ -0,0 +1,92 @@
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Configuration;
+
+namespace Microsoft.Identity.Web
+{
+    /// <summary>
+    ///  Scope authorization handler that need to be called for a specific requirement type.
+    ///  In this case, ScopeAuthorizationRequirement.
+    /// </summary>
+    public class ScopeAuthorizationHandler : AuthorizationHandler<ScopeAuthorizationRequirement>
+    {
+        private readonly IConfiguration _configuration;
+
+        /// <summary>
+        /// Constructor for the scope authorization handler, which takes a configuration.
+        /// </summary>
+        /// <param name="configuration">Configuration.</param>
+        public ScopeAuthorizationHandler(IConfiguration configuration)
+        {
+            _configuration = configuration;
+        }
+
+        /// <summary>
+        ///  Makes a decision if authorization is allowed based on a specific requirement.
+        /// </summary>
+        /// <param name="context">AuthorizationHandlerContext.</param>
+        /// <param name="requirement">Scope requirement.</param>
+        /// <returns>Task.</returns>
+        protected override Task HandleRequirementAsync(
+            AuthorizationHandlerContext context,
+            ScopeAuthorizationRequirement requirement)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            // The resource is either the HttpContext or the Endpoint directly when used with the
+            // authorization middleware
+            var endpoint = context.Resource switch
+            {
+                HttpContext httpContext => httpContext.GetEndpoint(),
+                Endpoint ep => ep,
+                _ => null,
+            };
+
+            var data = endpoint?.Metadata.GetMetadata<IAuthRequiredScopeMetadata>();
+
+            IEnumerable<string>? scopes = null;
+            if (requirement?.RequiredScopesConfigurationKey != null)
+            {
+                scopes = _configuration.GetValue<string>(requirement?.RequiredScopesConfigurationKey)?.Split(' ');
+            }
+
+            if (scopes == null)
+            {
+                scopes = requirement?.AllowedValues ?? data?.AcceptedScopes;
+            }
+
+            // Can't determine what to do without scope metadata, so proceed
+            if (scopes is null)
+            {
+                //context.Succeed(requirement);
+                return Task.CompletedTask;
+            }
+
+            Claim? scopeClaim = context.User.FindFirst(ClaimConstants.Scp) ?? context.User.FindFirst(ClaimConstants.Scope);
+
+            if (scopeClaim is null)
+            {
+                return Task.CompletedTask;
+            }
+
+            if (scopeClaim != null && scopeClaim.Value.Split(' ').Intersect(scopes).Any())
```

Comment:
Is there a way to preprocess the scopes once per endpoint + config "version" (via the reload token?) into something like a `HashSet<>` so that the `Intersect` call can be avoided? `Intersect` allocates and mutates a `HashSet<>` every time it it called.

Reply:
what is a reload token? can you point us to a doc?

Reply:
If you allow me to rephrase my more fundamental underlying point: consider pre-computing as much as is practical to minimize the work done on the critical path of handling a request.

----------------------------

A reload token is a way be notified if the underlying `IConfiguration` has changed. One can be retrieved from [`IConfiguration.GetReloadToken`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.iconfiguration.getreloadtoken?view=dotnet-plat-ext-5.0) and a callback can be registered using [`IChangeToken.RegisterChangeCallback`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.primitives.ichangetoken.registerchangecallback?view=dotnet-plat-ext-5.0#Microsoft_Extensions_Primitives_IChangeToken_RegisterChangeCallback_System_Action_System_Object__System_Object_).

Now, I have no idea if that's how a policy is supposed to monitor for config changes or not... I don't know this part of ASP.NET Core well.

-------------------------

⚠ WARNING ⚠<br>
Rampant speculation below here

_Briefly_ skimming the [Policy-based authorization in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/security/authorization/policies?view=aspnetcore-5.0) documentation, I _feels_ like the `ScopeAuthorizationRequirement` type is where the pre-processed data structures can live. (I don't know how endpoint metadata comes into play here... Maybe there need to be two places the pre-processed data is cached: per-endpoint if there's any endpoint customization and in the requirement itself?)

Here's a _possible_, **_untested_** implementation of caching the merged allowed values in the requirement. I've marked places where I'm unsure of things with "???". The existing `ScopeAuthorizationRequirement` didn't have access to an `IConfiguration` instance, so I added a helper internal method `GetMergedAllowedValues` to bind it to config at runtime. If the requirement could take a `IConfiguration` instance directly, the code would be a _ton_ simplier. But, I don't know how policies are supposed to be implemented. I'm winging this after skimming 5 minutes. :-) Intended use in the handler:

```
HashSet<string>? allowedScopes = requirement?.GetMergedAllowedValues(_configuration);
```

And the revised requirement:

```csharp
#nullable enable

namespace Microsoft.Identity.Web
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading;

    using Microsoft.AspNetCore.Authorization;
    using Microsoft.Extensions.Configuration;

    public class ScopeAuthorizationRequirement : IAuthorizationRequirement, IDisposable
    {
        private readonly List<string>? _allowedValues;

        private MergedAllowedValuesHelper? _mergedAllowedValuesHelper;

        public ScopeAuthorizationRequirement(IEnumerable<string>? allowedValues = null)
        {
            if (allowedValues is not null)
            {
                // make a copy, in case the passed IEnumerable can only be enumerated once
                _allowedValues = allowedValues.ToList();
            }
        }

        public ICollection<string>? AllowedValues => _allowedValues;

        public string? RequiredScopesConfigurationKey { get; init; }

        public void Dispose()
        {
            MergedAllowedValuesHelper? helper = Interlocked.Exchange(ref _mergedAllowedValuesHelper, null);
            helper?.Dispose();
        }

        internal HashSet<string> GetMergedAllowedValues(IConfiguration config)
        {
            // ??? Can this be replaced with a Lazy<MergedAllowedValuesHelper> somehow?

            if (_mergedAllowedValuesHelper is not null)
            {
                return _mergedAllowedValuesHelper.MergedAllowedValues;
            }

            var newMergedAllowedValuesHelper = new MergedAllowedValuesHelper(this, config);

            // Attempt to publish our instances as THE helper.
            MergedAllowedValuesHelper? oldMergedAllowedValuesHelper = Interlocked.CompareExchange(ref _mergedAllowedValuesHelper, newMergedAllowedValuesHelper, null);

            if (oldMergedAllowedValuesHelper is not null)
            {
                // We lost the race to publish. Clean up the junk we created
                // and return the value from the instance that was published.
                newMergedAllowedValuesHelper.Dispose();
                return oldMergedAllowedValuesHelper.MergedAllowedValues;
            }
            else
            {
                // We were first to publish. Return our value. That is also
                // what will be returned from now on.
                return newMergedAllowedValuesHelper.MergedAllowedValues;
            }
        }

        /// <summary>
        /// Handles merging the allowed values specificed on a given <see
        /// cref="ScopeAuthorizationRequirement"/> with any from config and
        /// monitoring for changes to that config.
        /// </summary>
        private class MergedAllowedValuesHelper : IDisposable
        {
            private readonly ScopeAuthorizationRequirement _parent;
            private readonly IConfiguration _config;
            private HashSet<string> _mergedAllowedValues;
            private IDisposable? _configChangeCallbackUnregisterer;

            public MergedAllowedValuesHelper(ScopeAuthorizationRequirement parent, IConfiguration config)
            {
                _parent = parent;
                _config = config;

                _mergedAllowedValues = MakeMergedAllowedValues();

                if (!string.IsNullOrEmpty(_parent.RequiredScopesConfigurationKey))
                {
                    // ??? There's a race between the call to
                    // MakeMergedAllowedValues, which may consult config, and
                    // the installation of this change callback. I don't know
                    // how to resolve that...
                    _configChangeCallbackUnregisterer = _config.GetReloadToken().RegisterChangeCallback(OnConfigChanged, this);
                }
            }

            public HashSet<string> MergedAllowedValues => _mergedAllowedValues;

            public void Dispose()
            {
                IDisposable? unregisterer = Interlocked.Exchange(ref _configChangeCallbackUnregisterer, null);
                unregisterer?.Dispose();
            }

            private static void OnConfigChanged(object thisAsObject)
            {
                ((MergedAllowedValuesHelper)thisAsObject).OnConfigChanged();
            }

            private HashSet<string> MakeMergedAllowedValues()
            {
                // ??? I assumed that config and ctor values are supposed to be merged, but that's not what the prior code did. Adjust as needed.
                string[] valuesFromConfig = Array.Empty<string>();

                if (!string.IsNullOrEmpty(_parent.RequiredScopesConfigurationKey))
                {
                    string? requiredScopesConfigValue = _config.GetValue<string>(_parent.RequiredScopesConfigurationKey);
                    if (requiredScopesConfigValue is not null)
                    {
                        valuesFromConfig = requiredScopesConfigValue.Split(' ');
                    }
                }

                var merged = new HashSet<string>((_parent._allowedValues?.Count ?? 0) + valuesFromConfig.Length);

                if (_parent._allowedValues is not null)
                {
                    foreach (string value in _parent._allowedValues)
                    {
                        merged.Add(value);
                    }
                }

                foreach (string value in valuesFromConfig)
                {
                    merged.Add(value);
                }

                return merged;
            }

            private void OnConfigChanged()
            {
                HashSet<string> newMergedAllowedValues = MakeMergedAllowedValues();

                // Atomically publish the new merged set.
                //
                // ??? This assumes that there will never be more than one
                // configuration changed callback in flight at once. I _think_
                // IConfiguration guarantees this...
                Interlocked.Exchange(ref _mergedAllowedValues, newMergedAllowedValues);
            }
        }
    }
}

```

Reply:
Yeah, I'd avoid going down change tokens and reloadable config unless you really have a need for that.

Reply:
ack. Thanks for the perspectives.